source/interface/azure_iot_mqtt.h

```diff
@@ -0,0 +1,376 @@
+
+    /**
+     * @brief Length of the client identifier.
+     */
+    uint16_t usClientIdentifierLength;
+
+    /**
+     * @brief MQTT user name. Set to NULL if not used.
+     */
+    const uint8_t * pcUserName;
+
+    /**
+     * @brief Length of MQTT user name. Set to 0 if not used.
+     *        or set pcUserName to NULL.
+     */
+    uint16_t usUserNameLength;
+
+    /**
+     * @brief MQTT password. Set to NULL if not used.
+     */
+    const uint8_t * pcPassword;
+
+    /**
+     * @brief Length of MQTT password. Set to 0 if not used
+     *        or set pcPassword to NULL.
+     */
+    uint16_t usPasswordLength;
+} AzureIoTMQTTConnectInfo_t;
+
+typedef struct AzureIoTMQTTSubscribeInfo
+{
+    /**
+     * @brief Quality of Service for subscription.
+     */
+    AzureIoTMQTTQoS_t xQOS;
+
+    /**
+     * @brief Topic filter to subscribe to.
+     */
+    const uint8_t * pcTopicFilter;
+
+    /**
+     * @brief Length of subscription topic filter.
+     */
+    uint16_t usTopicFilterLength;
+} AzureIoTMQTTSubscribeInfo_t;
+
+typedef struct AzureIoTMQTTPublishInfo
+{
+    /**
+     * @brief Quality of Service for message.
+     */
+    AzureIoTMQTTQoS_t xQOS;
+
+    /**
+     * @brief Whether this is a retained message.
+     */
+    uint8_t xRetain;
+
+    /**
+     * @brief Whether this is a duplicate publish message.
+     */
+    uint8_t xDup;
+
+    /**
+     * @brief Topic name on which the message is published.
+     */
+    const uint8_t * pcTopicName;
+
+    /**
+     * @brief Length of topic name.
+     */
+    uint16_t usTopicNameLength;
+
+    /**
+     * @brief Message payload.
+     */
+    const void * pvPayload;
+
+    /**
+     * @brief Message payload length.
+     */
+    size_t xPayloadLength;
+} AzureIoTMQTTPublishInfo_t;
+
+typedef struct AzureIoTMQTTDeserializedInfo
+{
+    uint16_t usPacketIdentifier;                 /** @brief Packet ID of deserialized packet. */
+    AzureIoTMQTTPublishInfo_t * pxPublishInfo;   /** @brief Pointer to deserialized publish info. */
+    AzureIoTMQTTResult_t xDeserializationResult; /** @brief Return code of deserialization. */
+} AzureIoTMQTTDeserializedInfo_t;
+
+typedef struct AzureIoTMQTTPacketInfo
+{
+    /**
+     * @brief Type of incoming MQTT packet.
+     */
+    uint8_t ucType;
+
+    /**
+     * @brief Remaining serialized data in the MQTT packet.
+     */
+    uint8_t * pucRemainingData;
+
+    /**
+     * @brief Length of remaining serialized data.
+     */
+    size_t xRemainingLength;
+} AzureIoTMQTTPacketInfo_t;
+
+typedef AzureIoTMQTT_t * AzureIoTMQTTHandle_t;
+
+typedef uint32_t ( * AzureIoTMQTTGetCurrentTimeFunc_t )( void );
+
+typedef void ( * AzureIoTMQTTEventCallback_t )( AzureIoTMQTTHandle_t pContext,
+                                                struct AzureIoTMQTTPacketInfo * pxPacketInfo,
+                                                struct AzureIoTMQTTDeserializedInfo * pxDeserializedInfo );
+
+
+/**
+ * @brief Initialize an AzureIoTMQTT context.
```

Comment:
Consider on the various doxygen blocks per function and even struct adding a note like (better word-smithed) "This function is for integration of the Azure IoT Middleware and the underlying MQTT stack and should in general not be invoked by applications directly."  

If a web search somehow landed me in the middle of this API block once we publish out doxygen and I was new to SDK, I'd worry it'd take me a bit to realize this wasn't the layer I needed to be at.